Add encoder support for drivetrain

VenusM3 opened the pull request:
## Added
* `getDistance()` to get the distance the robot passed since the last encoders reset.
* `resetEncoders()` to reset the encoders.
* `getLeftWheelsRPM()` to get the left wheels' RPM.
* `getRightWheelsRPM()` to get the right wheels' RPM.
* `getAverageWheelsRPM()` to get the average RPM of both sides' wheels.
* Some constants that are reused during calculation

Note - We need to change the constant kWheelCircumference to the actual value because thus far I couldn't get it, and we need to check if the motor to wheel rotation ratio is indeed 4:1 and edit kMotorToWheelRatio accordingly

Closes #16

guy-av commented on src/main/java/frc/robot/subsystems/DriveTrain.java:
```diff
@@ -9,6 +9,7 @@
 import edu.wpi.first.wpilibj.drive.DifferentialDrive;
 import edu.wpi.first.wpilibj.motorcontrol.MotorControllerGroup;
 import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.Constants;
```
Notice that this imports `Constants.*`, not `DriveTrainConstants.*` which is the one you want.
`DriveTrainConstants` is imported a line below.

guy-av commented on src/main/java/frc/robot/subsystems/DriveTrain.java:
```diff
@@ -22,6 +23,8 @@ public class DriveTrain extends SubsystemBase {
 
   private DifferentialDrive m_drive;
 
+  
+
```
Unnecessary, remove please.

guy-av commented on src/main/java/frc/robot/subsystems/DriveTrain.java:
```diff
@@ -41,4 +44,31 @@ public void arcadeDrive(double move, double rot) {
+  }
+
+  public double getAverageWheelsRPM() {
+    return (getLeftWheelsRPM() + getRightWheelsRPM()) / 4;
```
Why divide by **4**? It is true you have four encoders but you calculate average of only two values, one is the average RPM of the left wheels and the other is the average RPM of the right wheels.

VenusM3 replied:
Oops, at the time I probably thought of the motor to wheel ratio or idk, will fix

guy-av commented on src/main/java/frc/robot/subsystems/DriveTrain.java:
```diff
@@ -41,4 +44,31 @@ public void arcadeDrive(double move, double rot) {
+  }
+
+  public double getLeftWheelsRPM() {
+    return ((m_leftFront.getSelectedSensorVelocity() * Constants.DriveTrainConstants.k100msTo60sRatio) / 2048 + (m_leftRear.getSelectedSensorVelocity() * Constants.DriveTrainConstants.k100msTo60sRatio / 2048)) / 2 / 4;
```
There is a constant for encoder resolution (2048), use it.

Also, the magic **4**, it also has a constant, but do check with the builders that the motor to wheel ratio of the drivetrain's tough box is `4:1`.

~(Pssst... It's `DriveTrainConstants. kEncoderResolution `)~

guy-av commented on src/main/java/frc/robot/subsystems/DriveTrain.java:
```diff
@@ -41,4 +44,31 @@ public void arcadeDrive(double move, double rot) {
+  }
+
+  public double getRightWheelsRPM() {
+    return ((m_rightFront.getSelectedSensorVelocity() * Constants.DriveTrainConstants.k100msTo60sRatio) / 2048 + (m_rightRear.getSelectedSensorVelocity() * Constants.DriveTrainConstants.k100msTo60sRatio / 2048)) / 2 / 4;
```
Instead of doing the same calculation four different times across two different functions, maybe use an external function that only transforms `x [Raw / 100ms]` to `y [RPM]`.

Something like the following signature can match the function:
```java
private double rawSpeedToRPM(double rawSpeed) {
  // Do calculations
}
```

guy-av reviewed (changes requested):
Had some comments, please resolve them (Or take a shot at convincing me I'm wrong 😉).

VenusM3 commented:
> Had some comments, please resolve them (Or take a shot at convincing me I'm wrong 😉).

Working on it, thanks

VenusM3 commented:
> Had some comments, please resolve them (Or take a shot at convincing me I'm wrong 😉).

alr done the requested changes

guy-av commented on src/main/java/frc/robot/subsystems/DriveTrain.java:
```diff
@@ -60,15 +57,19 @@ public void resetEncoders() {
+  private double rawSpeedToRPM(double rawSpeed) {
+    return rawSpeed * DriveTrainConstants.k100msTo60sRatio / DriveTrainConstants.kEncoderResolution;
   }
 
```
Remove this empty line

guy-av commented on src/main/java/frc/robot/subsystems/DriveTrain.java:
```diff
@@ -60,15 +57,19 @@ public void resetEncoders() {
+  }
+
+  private double rawSpeedToRPM(double rawSpeed) {
+    return rawSpeed * DriveTrainConstants.k100msTo60sRatio / DriveTrainConstants.kEncoderResolution;
```
Great, this line gives motor RPM, but would you ever want the motor RPM? Usually the wheel RPM is what counts (Multiplying by 4 as in the `get...RPM()` functions.

VenusM3 replied:
Aren't we supposed to divide by 4? Because every 4 motor spins = 1 wheel spin

guy-av replied:
Yes, my bad! You should divide by four, not multiply.
But the point I was trying to make still stands:
This function is supposed to translate raw motor speed to RPM, but whose RPM? The motor? Or the wheel?
Is there a case when you would rather use the RPM of the motor itself on a drivetrain?
If not, you can divide by four inside the function, making the function translate raw speed to wheel RPM.

guy-av commented on src/main/java/frc/robot/subsystems/DriveTrain.java:
```diff
@@ -41,4 +41,35 @@ public void arcadeDrive(double move, double rot) {
   }
+
+  public double getDistance() {
+    // Calculates the average wheel sensor position(for example, if the right wheels did a one full rotation, and the left wheels did a 50% rotation, averageWheels will be 0.75 rotation)
+    double averageWheels = (m_leftFront.getSelectedSensorPosition() + m_leftRear.getSelectedSensorPosition() + m_rightFront.getSelectedSensorPosition() + m_rightRear.getSelectedSensorPosition()) / 4 / DriveTrainConstants.kMotorToWheelRatio / DriveTrainConstants.kEncoderResolution; 
```
This lines are way too complicated, suggesting separating calculations into several commands.

```suggestion
    double leftAvgDistance = (leftFront.get...() + leftRear.get...()) / 2;
    double rightAvgDistance = ...;  // same but for right
    double centralAvg = (leftAvg + rightAvg) / 2;
    // and here transform centralAvg to the units we want
```

guy-av reviewed (changes requested):
Some new comments!

guy-av commented:
Keep working normally for now but know this, when addressing comments and changing your PR it is usually better to alter the original commit the reviewer commented about. Let's say you have a commit adding a subsystem, and I comment about one of the functions, you don't want to add another commit fixing issues, we would rather no "broken" commits. So changing the original commit would spare you from additional ones.

This weekend or on Sunday we shall sit and learn how to alter commit history, please be careful if trying to do it on your own cause it isn't as easy as the internet makes it sounds. Either way, please don't `git push` an altered commit history before @Roy or myself make sure it is okay.

VenusM3 commented:
> Keep working normally for now but know this, when addressing comments and changing your PR it is usually better to alter the original commit the reviewer commented about. Let's say you have a commit adding a subsystem, and I comment about one of the functions, you don't want to add another commit fixing issues, we would rather no "broken" commits. So changing the original commit would spare you from additional ones.
> 
> This weekend or on Sunday we shall sit and learn how to alter commit history, please be careful if trying to do it on your own cause it isn't as easy as the internet makes it sounds. Either way, please don't `git push` an altered commit history before @roy or myself make sure it is okay.

Alright, currently I am sick(again, apparently viral...again, this time it's a voice box infection possibly), I hope I'll be able to come in Sunday, in the event I am not able to, is a Zoom/Google Meet/or any other type of call possible to talk about that?

guy-av commented:
> > Keep working normally for now but know this, when addressing comments and changing your PR it is usually better to alter the original commit the reviewer commented about. Let's say you have a commit adding a subsystem, and I comment about one of the functions, you don't want to add another commit fixing issues, we would rather no "broken" commits. So changing the original commit would spare you from additional ones.
> > This weekend or on Sunday we shall sit and learn how to alter commit history, please be careful if trying to do it on your own cause it isn't as easy as the internet makes it sounds. Either way, please don't `git push` an altered commit history before @roy or myself make sure it is okay.
> 
> Alright, currently I am sick(again, apparently viral...again, this time it's a voice box infection possibly), I hope I'll be able to come on Sunday, in the event I am not able to, is a Zoom/Google Meet/or any other type of call possible to talk about that?

Oh dear god, again?
Discord/Zoom/Google are all fine just let me know which you prefer and we'll schedule it.

VenusM3 commented:
> > > Keep working normally for now but know this, when addressing comments and changing your PR it is usually better to alter the original commit the reviewer commented about. Let's say you have a commit adding a subsystem, and I comment about one of the functions, you don't want to add another commit fixing issues, we would rather no "broken" commits. So changing the original commit would spare you from additional ones.
> > > This weekend or on Sunday we shall sit and learn how to alter commit history, please be careful if trying to do it on your own cause it isn't as easy as the internet makes it sounds. Either way, please don't `git push` an altered commit history before @roy or myself make sure it is okay.
> > 
> > 
> > Alright, currently I am sick(again, apparently viral...again, this time it's a voice box infection possibly), I hope I'll be able to come on Sunday, in the event I am not able to, is a Zoom/Google Meet/or any other type of call possible to talk about that?
> 
> Oh dear god, again? Discord/Zoom/Google are all fine just let me know which you prefer and we'll schedule it.

Yeah again, the doctor says it's fine and that it's viral. Discord is preferable, I don't mind talking in Sunday, assuming I won't have any escalation with the current symptoms I got

GlebiN commented:
seems legit? @royayal1

VenusM3 commented:
(hopefully yes 🙏)

royayal1 reviewed (dismissed):
seems legit
